vnext/Microsoft.ReactNative/JsiReader.cpp

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "JsiReader.h"
+#include <crash/verifyElseCrash.h>
+
+namespace winrt::Microsoft::ReactNative {
+
+//===========================================================================
+// JsiReader implementation
+//===========================================================================
+
+JsiReader::JsiReader(facebook::jsi::Runtime &runtime, const facebook::jsi::Value &root) noexcept
+    : m_runtime(runtime), m_root(root) {
+  SetValue(root);
+}
+
+JSValueType JsiReader::ValueType() noexcept {
+  if (m_currentPrimitiveValue) {
+    if (m_currentPrimitiveValue.value().isString()) {
+      return JSValueType::String;
+    } else if (m_currentPrimitiveValue.value().isBool()) {
+      return JSValueType::Boolean;
+    } else if (m_currentPrimitiveValue.value().isNumber()) {
+      double number = m_currentPrimitiveValue.value().getNumber();
+
+      // unfortunately JSI doesn't differentiate int and double
+      // here we test if the double value can be converted to int without data loss
+      // treat it like an int if we succeeded
+
+      if (floor(number) == number && MinSafeInteger <= number && number <= MaxSafeInteger) {
+        return JSValueType::Int64;
+      } else {
+        return JSValueType::Double;
+      }
+    }
+  } else if (m_nonPrimitiveValues.size() > 0) {
+    return m_nonPrimitiveValues[m_nonPrimitiveValues.size() - 1].Action == ContinuationAction::MoveToNextObjectProperty
+        ? JSValueType::Object
+        : JSValueType::Array;
+  }
+  return JSValueType::Null;
+}
+
+bool JsiReader::GetNextObjectProperty(hstring &propertyName) noexcept {
+  if (m_nonPrimitiveValues.size() == 0) {
+    return false;
+  }
+
+  auto &top = m_nonPrimitiveValues[m_nonPrimitiveValues.size() - 1];
+  if (top.Action != ContinuationAction::MoveToNextObjectProperty) {
+    return false;
+  }
+
+  top.Index++;
+  if (top.Index < static_cast<int>(top.PropertyNames.value().size(m_runtime))) {
```

Comment:
>if [](start = 2, length = 2)

I wonder if JSI has a more efficient way to iterate over properties.

Reply:
Unfortunately no, there is no iteration at all. `getPropertyNames` is the only API in `jsi::Object` to know what properties this object has.